runner/copy.go

```diff
@@ -75,28 +78,64 @@ func (m *MaterializedChunkRunner) Run() (types.CompletionWatcher, error) {
 			jobWatcher.EndWatch(fmt.Errorf("failed to create iterator: %w", err))
 			return
 		}
-		i := 0
-		for it.Next() {
-			i += 1
-			value := it.Value().Value
-			entity := it.Value().Entity
-			err := m.Table.Set(entity, value)
-			if err != nil {
-				jobWatcher.EndWatch(fmt.Errorf("could not set table: %w", err))
-				return
+		// Buffer the records channel to ensure readers won't block the writer loop below
+		ch := make(chan provider.ResourceRecord, resourceRecordBufferSize)
+		// Using a buffered error channel to capture any errors that may occur while trying to
+		// set values; buffering the error channel prevents the goroutines from blocking when
+		// trying to write to the channel.
+		errCh := make(chan error, 1)
+		var wg sync.WaitGroup
+		wg.Add(waitGroupSize)
+		// Create a "pool" of 100 goroutines to write to the table to better utilize CPU resources
+		// and speed up the write operations to the inference table.
+		for idx := 0; idx < waitGroupSize; idx++ {
+			go func() {
+				defer wg.Done()
+				for record := range ch {
+					if err := m.Table.Set(record.Entity, record.Value); err != nil {
+						select {
+						case errCh <- fmt.Errorf("could not set value to table: %w", err):
```

Comment:
because errCh has a buffer of 1, this case will only happen if no other errors are waiting to be processed. If another go-routine has an error, we will ignore this current error.

Reply:
This is OK, right? We want to fail on the first error, correct? If even one record fails to materialize, then we should consider the materialization a failure, right?